library/bigip_software_facts.py

```diff
@@ -0,0 +1,317 @@
+      filter: "version:12.1.1"
+  delegate_to: localhost
+'''
+
+RETURN = '''
+images:
+    description:
+        List of base image ISOs that are present on the unit.
+    returned: changed
+    type: list of dict
+    sample:
+        images:
+            - build: "0.0.184"
+              fileSize: "1997 MB",
+              lastModified: "Sun Oct  2 20:50:04 2016",
+              name: "BIGIP-12.1.1.0.0.184.iso",
+              product: "BIG-IP",
+              version: "12.1.1"
+hotfixes:
+    description:
+        List of hotfix ISOs that are present on the unit.
+    returned: changed
+    type: list of dict
+    sample:
+        hotfixes:
+            - build: "2.0.204"
+              fileSize: "1997 MB",
+              lastModified: "Sun Oct  2 20:50:04 2016",
+              name: "12.1.1-hf2.iso",
+              product: "BIG-IP",
+              version: "12.1.1"   
+volumes:     
+    description:
+        List the volumes present on device.
+    returned: changed
+    type: list of dict
+    sample:
+       volumes:
+            - basebuild: "0.0.184",
+              build: "0.0.184",
+              name: "HD1.2",
+              product: "BIG-IP",
+              status: "complete",
+              version: "12.1.1"              
+'''
+
+
+from ansible.module_utils.f5_utils import (
+    AnsibleF5Client,
+    AnsibleF5Parameters,
+    HAS_F5SDK,
+    F5ModuleError,
+)
+
+
+class Parameters(AnsibleF5Parameters):
+    returnables = [
+        'name', 'build', 'version', 'product',
+        'lastModified', 'basebuild', 'status',
+        'active', 'fileSize'
+        ]
+
+    @property
+    def include(self):
+        requested = self._values['include']
+        valid = ['all', 'volume', 'image', 'hotfix']
+
+        if not set(requested).issubset(set(valid)):
+            raise F5ModuleError(
+                'Include parameter may only be specified as one or more of the following: {0}'.format(', '.join(valid))
+            )
+
+        if 'all' in requested:
+            return ['all']
+        else:
+            return requested
+
+    @property
+    def filter(self):
+        requested = self._values['filter']
+        keys = ['name', 'build', 'version', 'status', 'active']
+        error = '"{0}" is not a valid filter format. Filters must have key:value format'.format(requested)
+        if requested is None:
+            return None
+
+        if ':' not in requested:
+            raise F5ModuleError(error)
+
+        key, value = requested.split(':')
+
+        if len(key) == 0 or len(value) == 0:
+            raise F5ModuleError(error)
+
+        if key not in keys:
+            raise F5ModuleError('"{0}" is not a supported filter. '
+                                'Supported key values are: {1}'.format(key, ', '.join(keys)))
+
+        return key, value
+
+
+class ModuleManager(object):
+    def __init__(self, client):
+        self.client = client
+
+    def exec_module(self):
+        manager = BaseManager(self.client)
```

Comment:
I think we missed the idea here. The base manager is what gets inherited by the concrete classes, it's not what is actually run.

What gets run should be the module manager which does the logic of which filter to use. So this Manager should really look more like the bigip_ssl_certificate ModuleManager where it gets a variety of managers and combines their results.